pkg/operator/provisioner.go

```diff
@@ -0,0 +1,207 @@
+*/
+package operator
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/kubernetes-incubator/external-storage/lib/controller"
+	"github.com/rook/rook/pkg/model"
+	rookclient "github.com/rook/rook/pkg/rook/client"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/client-go/pkg/api/v1"
+)
+
+type rookVolumeProvisioner struct {
+	clusterManager *clusterManager
+
+	// Configuration of rook volume provisioner
+	provConfig provisionerConfig
+}
+
+type provisionerConfig struct {
+	// Required:  the pool name to provision volumes from.
+	pool string
+
+	// Optional: Namespace of the cluster. Default is `rook`
+	clusterNamespace string
+
+	// Optional: Name of the cluster. Default is `rook`
+	clusterName string
+}
+
+func newRookVolumeProvisioner(clusterManager *clusterManager) controller.Provisioner {
+	return &rookVolumeProvisioner{
+		clusterManager: clusterManager,
+	}
+}
+
+// Provision creates a storage asset and returns a PV object representing it.
+func (p *rookVolumeProvisioner) Provision(options controller.VolumeOptions) (*v1.PersistentVolume, error) {
+
+	var err error
+	if options.PVC.Spec.Selector != nil {
+		return nil, fmt.Errorf("claim Selector is not supported")
+	}
+
+	glog.V(4).Infof("rookVolumeProvisioner: VolumeOptions %v", options)
+
+	cfg, err := parseClassParameters(options.Parameters)
+	if err != nil {
+		return nil, err
+	}
+	p.provConfig = *cfg
+
+	glog.V(4).Infof("rookVolumeProvisioner: creating volume with configuration %+v", p.provConfig)
+
+	capacity := options.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
+	requestBytes := capacity.Value()
+
+	imageName := fmt.Sprintf("kubernetes-dynamic-%s-%s", options.PVName, uuid.NewUUID())
+
+	rookClient, err := p.clusterManager.getRookClient(p.provConfig.clusterNamespace)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get rook client: %v", err)
+	}
+
+	res, err := createVolume(imageName, p.provConfig.pool, requestBytes, rookClient)
+	if err != nil {
+		return nil, err
+	}
+	glog.V(1).Infof("Rook block image created: %s", res)
+
+	rookClientInfo, err := rookClient.GetClientAccessInfo()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get rook client information: %v", err)
+	}
+	monitors := processMonAddresses(rookClientInfo.MonAddresses)
+	radosUser := fmt.Sprintf("%s-rbd-user", p.provConfig.clusterName)
+	secretRef := new(v1.LocalObjectReference)
+	secretRef.Name = fmt.Sprintf("%s-rbd-user", p.provConfig.clusterName)
+
+	pv := &v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: options.PVName,
+		},
+		Spec: v1.PersistentVolumeSpec{
+			PersistentVolumeReclaimPolicy: options.PersistentVolumeReclaimPolicy,
+			AccessModes:                   options.PVC.Spec.AccessModes,
+			Capacity: v1.ResourceList{
+				v1.ResourceName(v1.ResourceStorage): capacity,
+			},
+			PersistentVolumeSource: v1.PersistentVolumeSource{
+				RBD: &v1.RBDVolumeSource{
```

Comment:
We dont have a replacement. The idea is to enhance it so that it no longer depends on ceph-tools. But this is proving to be a bit difficult because the rbd plugin uses `rbd lock`. There are ways to do this but requires many patches and a really new kernel. I will create a issue detailing all the things we need.